CONTRIBUTING.adoc

```diff
@@ -54,194 +66,225 @@ However, to avoid doing too much work all the time, we have checked the generate
-* Comment if you want attention from someone (e.g. a re-review after
-changes). Github does not make it easy to signal this state otherwise,
-and people may not be notified if you just push commits.
-
-.Pre-submit checklist:
-* `$(nix-build default.nix -A dev.scripts.updateMaterialized)` to update the materialized Nix files.
-* `$(nix-shell shell.nix --run fix-stylish-haskell)` to fix any formatting issues.
-* `nix build -L -f ci.nix --arg supportedSystems '[ builtins.currentSystem ]'` to check that everything builds
-
-You can build some things with Nix locally if you like, but it's fine to let the CI do that for you.
-
-=== How to do code reviews
-
-.Guidelines for reviewing PRs:
-* Try to review PRs where your review is requested within a few days.
-This should be nearly-top-priority work.
-* If you don’t understand something then ask for an explanation.
-** For the author: this explanation should ideally be added as a comment - you’re
-going to write it anyway, and future readers are likely to be
-just as confused as the reviewer.
+For example:
 
-== Continuous integration
+----
+{- Note [How to write a note]
+A note should look a bit like this.
 
-We have one CI systems at the moment:
-- Hydra
-- Hercules (disabled, but will be back)
+Go wild, write lots of stuff!
 
-The CI will report statuses on your PRs with links to the logs in case of
-failure. Pull requests cannot be merged without the CI going green.
+Here's a small diagram:
+A ----> B >> C
 
-.Troubleshooting CI
-* Because the CI is not necessarily run on
-the merge commit that is created when the PR is merged, it is possible
-that merging a green PR can result in the CI being broken on master.
-This shouldn’t happen frequently, but be aware that it’s possible.
+And of course, you should see Note [Another note].
+-}
+----
 
-=== Hydra
+Notes are a great place to put substantial discussion that you need to refer to from multiple places.
+For example, if you used an encoding trick to fit more data into an output format,
+you could write a Note describing the trick (and justifying its usage!), and then refer to it from the encoder and the decoder.
+
+=== Formatting
+
+We use `stylish-haskell` for Haskell code formatting.
+CI checks that running `stylish-haskell` is a no-op, so if you don’t apply it them your PR will not go green.
+To avoid annoyance, set up your editor to run it automatically.
+The `nix-shell` environment provides a `stylish-haskell` binary of the correct version.
+
+You can run `stylish-haskell` over your tree using the `fix-stylish-haskell` script provided by the `nix-shell` environment,
+or directly by running `$(nix-shell shell.nix --run fix-stylish-haskell)`.
+
+=== Compiler warnings
+
+The CI builds Haskell code with `-Werror`, so will fail if there are any compiler warnings.
+So fix your own warnings!
+
+If the warnings are stupid, we can turn them off, e.g. sometimes it makes sense to add `-Wno-orphans` to a file where we know it's safe.
```

Comment:
(it's always safe, there's nothing wrong about using orphan instances in application code at all)

Reply:
@effectfully  you should be fighting for our orphan instances cause in the purescript community! they refuse to allow orphan instances 😭

Reply:
Not using orphan instances in application code is a stupid restriction that people argue against all the time: [1](https://www.reddit.com/r/haskell/comments/cyfs94/someones_haskell_disappointment_gist_i_came_across/eysjq9t/), [2](https://www.reddit.com/r/haskell/comments/cyfs94/someones_haskell_disappointment_gist_i_came_across/eyrvm94/), [3](https://www.reddit.com/r/haskell/comments/5rcfyd/haskell_maxims_and_arrows/dd6e7sd/)

And allowing orphan instances in general is way too useful to be prohibited completely: [1](https://mail.haskell.org/pipermail/libraries/2008-August/010399.html), [2](https://www.reddit.com/r/haskell/comments/ch0ye/are_orphan_instances_good_or_bad/)

It's like when people are afraid of enabling the `UndecidableInstances` extension, because there's the "Undecidable" word in it.

Reply:
I know Roman has strong feelings about orphans ;)